### PR TITLE
Fix staking reward issue

### PR DESCRIPTION
**Pull request by thiolliere**
Fix https://github.com/paritytech/substrate/issues/4121
Fix https://github.com/paritytech/substrate/issues/4220
Fix https://github.com/paritytech/substrate/issues/4331

### TODO

* [ ] maybe better naming
* [ ] maybe improvement on storage structure
* [x] Kusama transition
* [ ] Test storage upgrade transition with some real chain put_code upgrade

### DONE

when implementing `OnSessionEnding::on_session_ending`, we must end era if the session ending is the last one of the era, this era end will reward validator and clear some storage. but we also potentially plan a new era in the future using the argument `will_apply_at`.
This new era planned must freeze the validators information for this future era: validator set, validators exposures, commissions and slot_stake.

This new era in planned with the storage `EraStartSessionIndex` which store the session index on which the era start from the current era to the futurest planned.
Era information is stored in `ValidatorForEra` and `SlotStakeForEra`.

### CODE REVIEW

File: frame/staking/src/mock.rs [hunk 1]
```diff
@@ -443,8 +447,6 @@ pub fn advance_session() {
 }
 
 pub fn start_session(session_index: SessionIndex) {
-	// Compensate for session delay
-	let session_index = session_index + 1;
```
Comment (thiolliere) [inline frame/staking/src/mock.rs]: this is one reason why tests with start_session has been modified, it makes them clearer IMHO, also most tests don't use this but start_era instead.

File: frame/staking/src/mock.rs [hunk 2]
```diff
@@ -409,11 +413,11 @@ pub fn check_nominator_exposure(stash: u64) {
 	);
 }
 
-pub fn assert_is_stash(acc: u64) {
+pub fn assert_is_stash(acc: AccountId) {
```
Comment (gavofyork) [inline frame/staking/src/mock.rs]: pass as `&`

**Comment by gavofyork** [comment]
The staking/reward/exposure dataset could be huge and complicated. Have you analysed the various complexity here?

**Comment by gavofyork** [comment]
I would prefer this be done in the way I outline in #4407

**Pull request closed without merging**